webapp/src/ts/services/feedback.service.ts

```diff
@@ -43,6 +61,63 @@ export class FeedbackService {
       .filter(i => !!i);
   }
 
+  private shouldGenerateFeedback(level:string, message:string, exceptionMessage:string, exception?) {
+    if (this.FEEDBACK_LEVEL !== level) {
+      return false;
+    }
+
+    // requiring a valid error to be logged to avoid cascades of feedback docs
+    if (!message || !exception) {
+      return false;
+    }
+
+    // don't double-log errors as a basic infinite loop defense
+    if (this.lastErrorMessage === message || this.lastErrorMessage === exceptionMessage) {
+      return false;
+    }
+
+    message = message?.toLowerCase() || '';
+    exceptionMessage = exceptionMessage?.toLowerCase() || '';
+
+    const matchesNoFeedback = this.NO_FEEDBACK_MESSAGES.find((item:string|RegExp) => {
+      if (item instanceof RegExp) {
+        return item.test(message) || item.test(exceptionMessage);
+      }
+      return message.includes(item) || exceptionMessage.includes(item);
+    });
+
+    return !matchesNoFeedback;
+  }
+
+  private getExceptionMessage(exception) {
+    if (!exception) {
+      return;
+    }
+
+    if (typeof exception.error === 'string') {
+      return exception.error;
+    }
+
+    return exception.reason || exception.message;
+  }
+
+  private async generateFeedbackOnError(level, ...args) {
```

Comment:
For performance, let's do the `    if (this.FEEDBACK_LEVEL !== level) {` check first. 99% of log messages won't be errors and it saves us doing a bunch of other unnecessary checks.

```suggestion
  private async generateFeedbackOnError(level, ...args) {
    if (this.FEEDBACK_LEVEL !== level) {
      return;
    }
```